### PR TITLE
feat(host,provider-sdk): support provider config updates

### PR DESCRIPTION
**Pull request by vados-cosmonic**
## Feature or Problem

This PR introduces a scheme for supporting provider config updates by using the `CONFIGDATA_<lattice>` KV bucket. 

- When a provider makes a config bundle, it picks an ID (UUID v7, could add host ID)
- When the config bundle the host is maintaining updates, we push the value to the key in the KV bucket
- The host passes the key relevant to the config bundle to the provider at startup
- When the provider starts, `provider-sdk` listens independently on NATS for updates to the relevant key, and observes config updates that match it's bundle.
- The provider can implement `on_config_update()` to respond to config updates.

Resolves https://github.com/wasmCloud/wasmCloud/issues/1648

### CODE REVIEW

**Review by brooksmtownsend — CHANGES_REQUESTED** [review]
In general I think that this implements the functionality well, but I do have a bit of a concern here that now providers are connecting to JetStream directly whereas before they weren't. I think this might be going in the wrong direction when it comes to treating providers more like components, and may make a future WASI provider more difficult to implement.

What I envisioned for solving this issue was a similar method of updating, but using the `ConfigBundle` that's in the host that's specifically designed to watch the config bucket for changes in the specified keys, then publishing configuration updates to a handler the provider can implement to receive those updates. I don't think that would require too many changes to this PR, and it would solve the concern about cleaning up leftover keys in the CONFIGDATA bucket.

TL;DR, I think providers directly accessing JetStream isn't a pattern we should support, and we should keep the host as the only entity that can access that core piece of data for us.

What do you think @vados-cosmonic ? I can completely see the pros in terms of functionality of how it's implemented here, I just worry about what it means for the future of providers.

**Comment by vados-cosmonic** [comment]
Hey @brooksmtownsend I definitely get those concerns! We currently have Providers doing a bit of connecting to NATS themselves (and there's nothing to stop them from doing so), so I'm not sure I quite see the difference (maybe Jetstream is a much more fundamental functionality than regular NATS connections?) *but* I definitely think that we could formalize it a different way. 

I *think* that in the future WASI providers will either:

1. Get limited access to sockets
2. Have to use an explicit interface that is filled in/triggered by the host upon updates/etc

In both cases, the changes to `provider-sdk` to move from reading Jetstream directly to either allowing access to jetstream (i.e. the provider is still doing the network access), to doing the listening *on behalf of* the WASI provider isn't too disrupted by this current implementation. That said, I do see your concern!

I think the problem here is that we've done our best to sort of... skip formalizing a host <-> provider communication mechanism, we've left it to NATS (which we've benefitted from)...

What do you think would be a good communication process between hosts and providers that could work for this use case? I was avoiding regular NATS listen/send because chunking may be required, and we don't really have any other mechanisms AFAICT.

**Comment by thomastaylor312** [comment]
Hey @vados-cosmonic I think I get where you're coming from, but I also think some context might be helpful here along with some commentary on some of the reasoning. So @brooksmtownsend brought up `ConfigBundle` because I explicitly designed the `bundle.changed().await` method to enable the host to watch for config changes and then send it over a new NATS topic to a provider. This matches how we deliver everything else to a provider so we are following the same pattern. With that in mind, a couple of notes:

> I think the problem here is that we've done our best to sort of... skip formalizing a host <-> provider communication mechanism, we've left it to NATS (which we've benefitted from)...

The formalization has been the provider SDK, which by rule has used a NATS topic API. However, this goes back to the same rationale about why we removed putting links directly into the bucket from the ctl interface and changed it back to a topic. Acknowledging that, yes, NATS KV is just a stream of messages under the hood, you would never have something that uses an API elsewhere connect directly to a database and receive values from it. Case in point of this: if something besides the host was relying on reading link definitions from the KV store, that is now making your database an API. We're going to have to change how those are structured soon because they cause a race condition. If something was relying on those structures, then they would now be broken. This would greatly expand the work needed for backwards compatibility with the host. 

> What do you think would be a good communication process between hosts and providers that could work for this use case?

The last point leads me to the answer to this question: It should be over NATS as part of the "API contract" (which as you noted, we could probably formalize better) that let's a provider listen for incoming config updates that it can then do something with. I think this is a much cleaner API for two reasons: First is that where the data is coming from and how it is shaped in storage is decoupled from what the API looks like. Second, if we have a bug in how things are being transmitted (or changes in storage) we don't have to recompile all providers with a new SDK version. The host can change it and since it is an API, the SDK just goes on consuming it.

> I was avoiding regular NATS listen/send because chunking may be required, and we don't really have any other mechanisms AFAICT.

This was an interesting point I hadn't thought of before, but I don't think it is a problem right now. This would mean that your config is over 1MB. We probably should be setting a max size on the config bucket when we set it up (likely 1MB), so we should open an issue for that. However, I don't think this is a problem in the short term

**Comment by vados-cosmonic** [comment]
Thanks @thomastaylor312 for the deeper explanation -- I think I see what ya'll were expecting the solution to look like with the current API conventions.

As far as the other stuff, that all sounds reasonable -- I'll refactor this PR to use a new NATS topic.

> This was an interesting point I hadn't thought of before, but I don't think it is a problem right now. This would mean that your config is over 1MB. We probably should be setting a max size on the config bucket when we set it up (likely 1MB), so we should open an issue for that. However, I don't think this is a problem in the short term

Well this certainly solves my reason to use KV directly -- will do this, and we can cross that bridge when someone tries to send too much config data.

File: crates/host/src/wasmbus/mod.rs [hunk 1]
```diff
@@ -317,13 +320,17 @@ impl wrpc_transport::Serve for WrpcServer {
     }
 }
 
+/// A (usually running) Provider instance
```
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply crates/host/src/wasmbus/mod.rs:323]: Yeah the provider could actually exit -- just having one of these doesn't necessary mean that a provider is running... For example we don't hold the actual `Child`... I'll update the comment but one thing that I noticed is that we actually just let the `Child` object drop

File: crates/host/src/wasmbus/config.rs [hunk 2]
```diff
@@ -349,7 +357,7 @@ mod tests {
         let (bar_tx, bar_rx) = watch::channel(HashMap::new());
         let (baz_tx, baz_rx) = watch::channel(HashMap::new());
 
-        let mut bundle = ConfigBundle::new(vec![
+        let mut bundle = ConfigBundle(vec![
```
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply crates/host/src/wasmbus/config.rs:360]: No it's not! the `::new` should be there!

File: crates/host/src/wasmbus/mod.rs [hunk 3]
```diff
@@ -2442,7 +2506,8 @@ impl Host {
                 "provider did not gracefully shut down in time, shutting down forcefully"
             );
         }
-        child.abort();
+        health_check_task.abort();
+        config_update_task.abort();
```
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply crates/host/src/wasmbus/mod.rs:2510]: So we didn't do that up until now, but I'd suggest addressing this in a follow up -- we actually *don't save* the `Child` task of the provider so we can try to do anything with it... We just :crossed_fingers: that the shutdown is processed correctly (which means the provider should shut itself down)...

File: crates/host/src/wasmbus/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply crates/host/src/wasmbus/mod.rs]: Updated the comment here!

File: crates/host/src/wasmbus/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply crates/host/src/wasmbus/mod.rs:2510]: Created an issue for this:

https://github.com/wasmCloud/wasmCloud/issues/2698

**Review by brooksmtownsend — APPROVED** [review]
Great stuff @vados-cosmonic !
